### PR TITLE
Update /data

### PR DESCRIPTION
**Pull request by Spandamn**
This will allow some really cool stuff like custom /data commands and:

Also gonna call `getDataPokemonHTML` in the Other Metas Chatplugin to improve its display.

### CODE REVIEW

**Comment by urkerab** [comment]
FYI I had a slightly different approach: https://hastebin.com/raw/mazijihaza

 - Created a single function that handles Pokémon, Items, Moves, Abilities and Natures
 - Put the function in `Chat` rather then `Tools`
 - Haven't finished code for handling alternate Pokémon icons (`template.spritenum`)
 - Tweaked `getNature` to include the description
 - Tweaked `dataSearch` to return the found thing to avoid looking it up again
 - Diff doesn't include code that looks up the mod or calls `onModifyPokemon`

**Comment by Zarel** [comment]
This code definitely belongs in `Chat` rather than `Tools`. Too bad about the code duplication.

**Comment by MacChaeger** [comment]
Ok you should really consider making a pull request to your pull request so you can do all those commits there and then commit that branch into this one, so you don't spam dev with updates.

**Comment by panpawn** [comment]
>making a pull request to your pull request

...?

But yeah, if you'd like help with Git, I can show you a few cool tricks if you'd like - it might make this kind of process a little smoother.

**Comment by Spandamn** [comment]
He prolly means making a pull request on the branch for that PR (patch-xx for this one)

**Comment by Spandamn** [comment]
Added a nice lil check for Generation:

**Comment by Spandamn** [comment]
Bump @Zarel 
(idk if we're allowed to bump PRs)

**Comment by Zarel** [comment]
Bumping is okay

**Comment by urkerab** [comment]
Well, my `template.spritenum` hack seems to work, I guess `/data` doesn't have to worry about all the little details that you can set in the Teambuilder. Also my past gen code works slightly differently so trying to do `/data Silvally, gen6` gives you the data for Rivalry instead, which I thought was cute.

**Comment by Spandamn** [comment]
That is a good idea. I'll make those changes in a while

**Comment by Spandamn** [comment]
data
Works now. Also tweaked the description for /data and /details a bit.

File: chat.js [hunk 1]
```diff
@@ -948,3 +948,409 @@ Chat.toDurationString = function (number, options) {
 	}
 	return parts.slice(positiveIndex).reverse().map((value, index) => value ? value + " " + unitNames[index] + (value > 1 ? "s" : "") : "").reverse().slice(0, precision).join(" ").trim();
 };
+
+Chat.getPokemonIcon = function (pokemon, facingLeft) {
```
Comment (QuiteQuiet) [inline chat.js]: Is there really not a better way to do this beside duplicating this function on the server as well?

Comment (Spandamn) [inline chat.js]: Well, like urkerab did, one can modify pokedex.js and make a new `template.spritenum` property.

Comment (Zarel) [inline chat.js]: My proposal was to add a pokemon/item icon tag that the client could translate

Comment (urkerab) [inline chat.js]: FYI my spritenum approach looked like this: https://hastebin.com/raw/cuholecogu

File: chat-plugins/info.js [hunk 2]
```diff
@@ -338,6 +338,9 @@ exports.commands = {
 		if (sep[1] && toId(sep[1]) in Tools.dexes) {
 			mod = Tools.mod(toId(sep[1]));
 		}
+		else if (sep[1] && Tools.getFormat(sep[1]).mod) {
```
Comment (panpawn) [inline chat-plugins/info.js]: I think PS convention for this would be to have this on the same line as the one above it

File: chat.js (hunk 1 above)
Comment (Spandamn) [inline chat.js]: ^ this is what seems great to me.

**Comment by Spandamn** [comment]
uhh damn.

**Comment by KrisOnline** [comment]
More like uhh span-damn xdddddddududucid

**Comment by panpawn** [comment]
I think I've said this before, but if you'd like help with rebasing this or squashing or anything like that, feel free to shoot me a PM on PS :)

**Comment by Spandamn** [comment]
Squashed and Rebased, should be fine now.
(Sorry I did this after so long, I didnt have access to my PC)
@Zarel

**Comment by Spandamn** [comment]
Past
Like CZ suggested, i added a check to remove stuff that doesn't exist in the mod's gen.

**Comment by Spandamn** [comment]
one-month-later bumping
@Zarel

**Comment by Zarel** [comment]
You've been working on it so I assumed it was unfinished.

**Comment by Zarel** [comment]
This is probably first blocked by a lack of syntax for item/pokemon icons. That'll need work first.

**Comment by Spandamn** [comment]
I'm sorry, I didn't quite get you...

**Comment by Zarel** [comment]
Short version: Give us more time

**Comment by Spandamn** [comment]
Okay

**Comment by Spandamn** [comment]
Big thanks to @HoeenCoder for helping me out :D

**Comment by Spandamn** [comment]
@Zarel, I dont really know whats blocking this... is it anything that I could help with?

**Comment by Zarel** [comment]
Um, a client pullreq to improve `sanitizeHTML` by replacing `<pokemonicon id="[pokemonid]" />` and `<itemicon id="[itemid]" />` with actual icons would be a good first step.

**Comment by Zarel** [comment]
Well, icons have been implemented now.

https://github.com/Zarel/Pokemon-Showdown-Client/commit/346bf935db3490a4b7e336863df58f134c7981b0

You should use these.

**Comment by Spandamn** [comment]
Fuck.

**Comment by Spandamn** [comment]
Will fix this in a while

**Comment by Zarel** [comment]
That is definitely not the right way to rebase a thing...

**Comment by panpawn** [comment]
If you need help rebasing, just let me know

**Comment by Zarel** [comment]
I tried to rebase it, but this repo's gotten seriously screwed up. :/

**Comment by panpawn** [comment]
You're probably going to have to reset this to the latest commit on main's master branch and then re-add/do your commits; simple enough.

Again, if you need help, feel free to PM me.

**Comment by Zarel** [comment]
I mean, you can fish for the old commits in the reflog, I guess?

**Comment by Zarel** [comment]
Would need to be him, though; I can't fish from his reflog.

**Comment by Zarel** [comment]
Hm, easiest solution is probably to cherry-pick.

**Comment by Zarel** [comment]
Here, I rebased with a cherry-pick.

**Comment by panpawn** [comment]
That works

File: chat-plugins/info.js [hunk 3]
```diff
@@ -388,19 +396,31 @@ exports.commands = {
...
 				} else {
-					buffer += '|c|~|/data-' + newTargets[i].searchType + ' ' + newTargets[i].name + '\n';
+					let sType = newTargets[i].searchType.charAt(0).toUpperCase() + newTargets[i].searchType.substring(1, newTargets[i].searchType.length);
+					if (Chat[`getData${sType}HTML`]) {
+						let template = mod["get" + (sType === "Pokemon" ? "Template" : sType)](newTargets[i].name);
+						if (sType === "Pokemon") {
+							if (mod.gen === 1) {
+								template.baseStats["spc"] = template.baseStats.spa;
+							}
+							if (mod.gen < 3) delete template.abilities;
```
Comment (Zarel) [inline chat-plugins/info.js]: Ugh, please stop mutating the outputs of dex data functions. This could easily cause other code to crash.

**Comment by Zarel** [comment]
Okay, now it's good. :p

**Pull request merged**